File: templates/problems/04_stack/02_min_stack/python/solution.py
class SolveMinStack:
    """Min stack supporting push, pop, top, and getMin in O(1).

    Time: O(1) per operation, Space: O(n)
    """

    def __init__(self) -> None:
        self.stack: list[int] = []
        self.min_stack: list[int] = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        top = self.stack.pop()
        if top == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def get_min(self) -> int:
        return self.min_stack[-1]

File: templates/problems/04_stack/02_min_stack/typescript/solution.ts
/**
 * Min stack supporting push, pop, top, and getMin in O(1).
 * Time: O(1) per operation, Space: O(n)
 */
export class SolveMinStack {
  private stack: number[] = [];
  private minStack: number[] = [];

  push(val: number): void {
    this.stack.push(val);
    if (this.minStack.length === 0 || val <= this.minStack[this.minStack.length - 1]) {
      this.minStack.push(val);
    }
  }

  pop(): void {
    const top = this.stack.pop()!;
    if (top === this.minStack[this.minStack.length - 1]) {
      this.minStack.pop();
    }
  }

  top(): number {
    return this.stack[this.stack.length - 1];
  }

  getMin(): number {
    return this.minStack[this.minStack.length - 1];
  }
}
